cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ProgramNotificationSubscriberService.java

```diff
@@ -263,27 +319,67 @@ private void handleProgramEvent(ProgramRunId programRunId, ProgramRunStatus prog
+            // STARTING => STARTING : state transition after app-fabric restart
+            LOG.debug("Ignoring unexpected request to transition program run {} from {} state to program " +
+                        "STARTING state.", programRunId, runRecordDetail.getStatus());
+            return;
+          }
+        } catch (IllegalStateException ex) {
+          LOG.error("Request to transition program run {} from non-existent state to program STARTING state " +
+                      "but multiple run IDs exist.", programRunId);
+        }
+
         String systemArgumentsString = properties.get(ProgramOptionConstants.SYSTEM_OVERRIDES);
         Map<String, String> systemArguments = systemArgumentsString == null ?
           Collections.emptyMap() : GSON.fromJson(systemArgumentsString, STRING_STRING_MAP);
         boolean isInWorkflow = systemArguments.containsKey(ProgramOptionConstants.WORKFLOW_NAME);
         boolean skipProvisioning = Boolean.parseBoolean(systemArguments.get(ProgramOptionConstants.SKIP_PROVISIONING));
+
+        ProgramOptions prgOptions = ProgramOptions.fromNotification(notification, GSON);
+        ProgramDescriptor prgDescriptor =
+          GSON.fromJson(properties.get(ProgramOptionConstants.PROGRAM_DESCRIPTOR), ProgramDescriptor.class);
+
         // if this is a preview run or a program within a workflow, we don't actually need to provision a cluster
         // instead, we skip forward past the provisioning and provisioned states and go straight to starting.
         if (isInWorkflow || skipProvisioning) {
-          ProgramOptions programOptions = ProgramOptions.fromNotification(notification, GSON);
-          ProgramDescriptor programDescriptor =
-            GSON.fromJson(properties.get(ProgramOptionConstants.PROGRAM_DESCRIPTOR), ProgramDescriptor.class);
-          appMetadataStore.recordProgramProvisioning(programRunId, programOptions.getUserArguments().asMap(),
-                                                     programOptions.getArguments().asMap(), messageIdBytes,
-                                                     programDescriptor.getArtifactId().toApiArtifactId());
+          appMetadataStore.recordProgramProvisioning(programRunId, prgOptions.getUserArguments().asMap(),
+                                                     prgOptions.getArguments().asMap(), messageIdBytes,
+                                                     prgDescriptor.getArtifactId().toApiArtifactId());
           appMetadataStore.recordProgramProvisioned(programRunId, 0, messageIdBytes);
         }
+
         recordedRunRecord = appMetadataStore.recordProgramStart(programRunId, twillRunId,
                                                                 systemArguments, messageIdBytes);
         writeToHeartBeatTable(recordedRunRecord,
                               RunIds.getTime(programRunId.getRun(), TimeUnit.SECONDS),
                               programHeartbeatTable);
+
+        if (isInWorkflow || skipProvisioning) {
```

Comment:
do we always want to call `appMetadataStore.recordProgramStart` even before calling `startInternal`?

Reply:
That is safe. If app-fabric restarts even before startInternal finishes successfully, upon restart, we will publish another STARTING message which results in calling startInternal again upon receiving the second STARTING message.